airbyte-integrations/connectors/source-pinterest/source_pinterest/components/auth.py

```diff
@@ -0,0 +1,63 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from base64 import standard_b64encode
+from dataclasses import dataclass
+from typing import Any, Mapping
+
+import backoff
+import requests
+from airbyte_cdk.models import FailureType
+from airbyte_cdk.sources.declarative.auth import DeclarativeOauth2Authenticator
+from airbyte_cdk.sources.streams.http.exceptions import DefaultBackoffException
+from airbyte_cdk.utils import AirbyteTracedException
+from airbyte_cdk.utils.airbyte_secrets_utils import add_to_secrets
+
+logger = logging.getLogger("airbyte")
+
+
+@dataclass
+class PinterestOauthAuthenticator(DeclarativeOauth2Authenticator):
+    def get_refresh_access_token_headers(self) -> Mapping[str, Any]:
+        creds_encoded = standard_b64encode(f"{self.get_client_id()}:{self.get_client_secret()}".encode("ascii")).decode("ascii")
+        return {"Authorization": f"Basic {creds_encoded}"}
+
+    @backoff.on_exception(
+        backoff.expo,
+        DefaultBackoffException,
+        on_backoff=lambda details: logger.info(
+            f"Caught retryable error after {details['tries']} tries. Waiting {details['wait']} seconds then retrying..."
+        ),
+        max_time=300,
+    )
+    def _get_refresh_access_token_response(self) -> Any:
+        try:
+            response = requests.request(
+                method="POST",
+                url=self.get_token_refresh_endpoint(),
+                headers=self.get_refresh_access_token_headers(),
+                data=self.build_refresh_request_body(),
+            )
+            if response.ok:
+                response_json = response.json()
+                access_key = response_json.get(self.get_access_token_name())
+                if not access_key:
+                    raise Exception(f"Token refresh API response was missing access token {self.get_access_token_name()}")
```

Comment:
what scenario would cause this to happen? This seems like suspicious behavior that we get a 200 OK, but the response is missing important information. It seems like this is expected since we are explicitly catching this. I think pending what the cause is, we should still raise some sort of TracedException here as well

Reply:
Thanks for the update. I'm still a bit curious, so the customer can have invalid pinterest creds, which will still make a valid request. But that response just won't contain any `access_key`. I just want to make sure we're providing the correct action item to the user to fix by understanding the root cause of why it happens

Reply:
I think this could be helpful for cases when the source returns status code `200` even if the access key cannot be provided due to incorrect credentials provided

Reply:
i see. yeah i think we can leave it in, but I do think its not good design that their API gives us a 200 when we effectively gave invalid input 😠

Reply:
Totally agree!